### PR TITLE
Add complex column for "scalar" complex data

### PR DESCRIPTION
**Pull request by jenshnielsen**
Complex array data types are already handled correctly. 

This is one of the possible solutions. Store complex numbers in a blob for each row similar to how arrays are stored. 

Possible alternatives.

* A complex row type only for complex values
* An interface that maps 1 complex number to 2 columns automatically

@QCoDeS/core What do you think should we go this way or try an alternative solution?

### CODE REVIEW

File: qcodes/dataset/sqlite_base.py [hunk 1]
```diff
@@ -246,7 +246,12 @@ def _convert_numeric(value: bytes) -> Union[float, int, str]:
         # if the reason was the conversion to float, and, if so, we are sure
         # that we need to return a string
         if "could not convert string to float" in str(e):
```
Comment (jenshnielsen) [inline qcodes/dataset/sqlite_base.py]: This logic is somewhat convoluted and the biggest downside of this solution

**Comment by jenshnielsen** [comment]
@WilliamHPNielsen and I had an offline discussion and agreed that it's better with a specific complex column datatype

File: qcodes/dataset/measurements.py [hunk 2]
```diff
@@ -179,6 +181,8 @@ def add_result(self, *res_tuple: res_type) -> None:
                 value = np.atleast_1d(value)
                 array_size = len(value.ravel())
                 if param_spec.type != 'array' and array_size > 1:
+                    # TODO this does not check that the dtype matches the sqlite
+                    # column type
```
Comment (jenshnielsen) [inline qcodes/dataset/measurements.py]: This is not a bug in this pr but does never the less need to be addressed

File: qcodes/dataset/sqlite_base.py [hunk 3]
```diff
@@ -271,6 +276,13 @@ def _adapt_float(fl: float) -> Union[float, str]:
     return float(fl)
 
 
+def _adapt_complex(value: Union[complex, np.complex]) -> sqlite3.Binary:
```
Comment (astafan8) [inline qcodes/dataset/sqlite_base.py]: Is this type annotation complete?

also, shall we make a `Complex` type as `Union[complex, np.complex, np.complex64, ...]`?

**Comment by jenshnielsen** [comment]
> 
> 
> Is the idea that 'complex' is sort of the same as 'numeric' (as in not 'array') but just represents complex numbers? if so, are there any complications for 'array' of complex numbers that we need to care about (i can think of only updating docs everywhere saying "for storing complex arrays (not scalars) just use 'array' type")?

Yes that is a separate task to add tests and docs for complex array parameters (which already seems to work)

File: qcodes/dataset/sqlite_base.py (hunk 3 above)
Comment (jenshnielsen) [reply qcodes/dataset/sqlite_base.py]: Yes

File: qcodes/utils/validators.py [hunk 4]
```diff
@@ -301,6 +301,28 @@ def validate(self, value: numbertypes, context: str = '') -> None:
...
+    validtypes = (complex, np.complex, np.complex64, np.complex128)
+
+    def __init__(self) -> None:
+
+        self._valid_values = ((1+1j), )
+
+    def validate(self, value: numbertypes, context: str = '') -> None:
+        if not isinstance(value, self.validtypes):
+            raise TypeError(
+                '{} is not an int or float; {}'.format(repr(value), context))
```
Comment (astafan8) [inline qcodes/utils/validators.py]:
```suggestion
                '{} is not complex; {}'.format(repr(value), context))
```

File: qcodes/utils/validators.py [hunk 5]
```diff
@@ -301,6 +301,28 @@ def validate(self, value: numbertypes, context: str = '') -> None:
         super().validate(castvalue, context=context)
 
 
+class ComplexNum(Validator):
```
Comment (jenshnielsen) [inline qcodes/utils/validators.py]: The validator needs a test

File: qcodes/dataset/measurements.py [hunk 6]
```diff
@@ -190,17 +194,23 @@ def add_result(self, *res_tuple: res_type) -> None:
...
                                      f"because its type class is "
-                                     f"'text', not 'numeric'.")
+                                     f"{param_spec.type!r}, not 'numeric'.")
             elif isinstance(value, str):
                 if param_spec.type != 'text':
                     raise ValueError(f"It is not possible to save a string "
                                      f"value for parameter {paramstr!r} "
                                      f"because its type class is "
                                      f"{param_spec.type!r}, not 'text'.")
+            elif isinstance(value, (complex, np.complex, np.complex128, np.complex64)):
```
Comment (astafan8) [inline qcodes/dataset/measurements.py]: shall we define `complex_types = (complex, np.complex, np.complex128, np.complex64)` in, say, `qcodes.math.<some module>` and re-use them everywhere in the code? (i actually wonder why numpy does not have such a definition in the library for the convenience of the humanity...)

File: qcodes/tests/dataset/test_measurement_context_manager.py [hunk 7]
```diff
@@ -1545,6 +1574,66 @@ def test_datasaver_2d_multi_parameters_array(channel_array_instrument):
...
+
+    with meas.run() as datasaver:
+        setparam.set(0)
+        with pytest.raises(ValueError, match=expected_msg):
+            datasaver.add_result((setparam, setparam()),
+                                 (param, param()))
+
+
+@pytest.mark.usefixtures("experiment")
+def test_save_complex_num_as_complex_raises(complex_num_instrument):
```
Comment (astafan8) [inline qcodes/tests/dataset/test_measurement_context_manager.py]:
```suggestion
def test_save_numeric_as_complex_raises(complex_num_instrument):
```

File: qcodes/dataset/measurements.py [hunk 8]
```diff
@@ -190,17 +194,23 @@ def add_result(self, *res_tuple: res_type) -> None:
...
-                if param_spec.type == 'text':
+                if param_spec.type != 'numeric':
                     raise ValueError(f"It is not possible to save a numeric "
                                      f"value for parameter {paramstr!r} "
                                      f"because its type class is "
-                                     f"'text', not 'numeric'.")
+                                     f"{param_spec.type!r}, not 'numeric'.")
             elif isinstance(value, str):
                 if param_spec.type != 'text':
                     raise ValueError(f"It is not possible to save a string "
```
Comment (astafan8) [inline qcodes/dataset/measurements.py]: just wondering, do we have parametrization on existing param_types for tests for these exceptions? if so, are we sure that the tests are executed also for the new 'complex' paramtype ?

Comment (jenshnielsen) [reply qcodes/dataset/measurements.py]: I think we should completely rewrite this logic

**Comment by jenshnielsen** [comment]
@WilliamHPNielsen It would be great to look at this together to make sure that I understand the refactoring that you did correctly

File: qcodes/dataset/measurements.py (hunk 6 above)
Comment (jenshnielsen) [reply qcodes/dataset/measurements.py]: Moved this to utils and prepared for the other complex pr

Comment (astafan8) [reply qcodes/dataset/measurements.py]: thanks!

**Comment by WilliamHPNielsen** [comment]
It looks good and correct, but perhaps we could add tests where we save a complex parameter alongside an array and where we save a list of complex numbers (although I hate that behaviour, it should be supported for now). Just to be sure that complex number do indeed work as scalars and strings in terms of the unraveling/raveling logic.

File: qcodes/dataset/measurements.py [hunk 9]
```diff
@@ -411,6 +414,8 @@ def reshaper(val, paramtype):
                 return float(val)
             elif paramtype == 'text':
                 return str(val)
+            elif paramtype == 'complex':
+                return complex(val)
```
Comment (jenshnielsen) [inline qcodes/dataset/measurements.py]: Even if this is perhaps a bit pathological it should have a test

Comment (astafan8) [reply qcodes/dataset/measurements.py]: ❤️

File: qcodes/utils/validators.py [hunk 10]
```diff
@@ -301,6 +303,28 @@ def validate(self, value: numbertypes, context: str = '') -> None:
...
```
Comment (jenshnielsen) [inline qcodes/utils/validators.py]: Should this be renames ComplexNumbers for consistency

Comment (astafan8) [reply qcodes/utils/validators.py]: yes :)

File: qcodes/dataset/measurements.py [hunk 11]
```diff
@@ -47,9 +47,11 @@ class ParameterTypeError(Exception):
 
 def is_number(thing: Any) -> bool:
     """
-    Test if an object can be converted to a number UNLESS it is a string
+    Test if an object can be converted to a float UNLESS it is a string or
+    complex.
```
Comment (jenshnielsen) [inline qcodes/dataset/measurements.py]: This is actually unused by now. After #1477. I will deprecate in another pr

**Comment by jenshnielsen** [comment]
@QCoDeS/core I think this is now functionally complete. I don't like how many places this hooks into the api but I believe it is correct and has sufficient tests

File: qcodes/dataset/sqlite_base.py [hunk 12]
```diff
@@ -1350,10 +1368,13 @@ def get_parameter_data(conn: ConnectionPlus,
 
         # if we have array type parameters expand all other parameters
         # to arrays
-        if 'array' in types and ('numeric' in types or 'text' in types):
+        if 'array' in types and ('numeric' in types or 'text' in types
+                                 or 'complex in types'):
```
Comment (astafan8) [inline qcodes/dataset/sqlite_base.py]: Looks like typo. Is it caught by any tests? what kind of tests would catch this?
```suggestion
                                 or 'complex' in types):
```

File: qcodes/utils/validators.py [hunk 13]
```diff
@@ -301,6 +303,28 @@ def validate(self, value: numbertypes, context: str = '') -> None:
...
+    def __init__(self) -> None:
+
+        self._valid_values = ((1+1j), )
+
+    def validate(self, value: numbertypes, context: str = '') -> None:
+        if not isinstance(value, self.validtypes):
+            raise TypeError(
+                '{} is not complex; {}'.format(repr(value), context))
+
+    is_numeric = False  # there is on meaningful way to sweep a complex number
```
Comment (astafan8) [inline qcodes/utils/validators.py]:
```suggestion
    is_numeric = False  # there is no meaningful way to sweep a complex number
```

File: qcodes/dataset/sqlite_base.py [hunk 14]
```diff
@@ -24,13 +24,15 @@
...
+from qcodes.utils.types import complex_types
+
 
 log = logging.getLogger(__name__)
 
 # represent the type of  data we can/want map to sqlite column
 VALUE = Union[str, Number, List, ndarray, bool]
 VALUES = List[VALUE]
-
+complex_type_union = Union[complex, np.complex, np.complex64, np.complex128]
```
Comment (astafan8) [inline qcodes/dataset/sqlite_base.py]: shall this be moved to `qcodes/utils/types.py`?

File: qcodes/dataset/sqlite_base.py (hunk 12 above)
Comment (jenshnielsen) [reply qcodes/dataset/sqlite_base.py]: As I understand it we cannot write a test that would catch this. The non empty string always evaluates to True. However removing the the `and ...` part of the if statement will only result in a slight performance degradation so there is no way of testing for it

Comment (astafan8) [reply qcodes/dataset/sqlite_base.py]: exactly, i got similar conclusions as well ((

anyway, this is yet another example that confirms that "code review" is important and necessary for ensuring code quality blablablabla.

**Pull request merged**